src/elf/symver.rs

```diff
@@ -0,0 +1,902 @@
+                )?
+            };
+
+            // Get a slice of bytes of the `version definition` section content.
+            let bytes: &'a [u8] = bytes.pread_with(offset, size)?;
+
+            Ok(Some(VerdefSection {
+                verstr,
+                bytes,
+                count,
+                ctx,
+            }))
+        }
+
+        /// Get an iterator over the [`Verdef`] entries.
+        #[inline]
+        pub fn iter(&'a self) -> VerdefIter<'a> {
+            self.into_iter()
+        }
+    }
+
+    impl<'a> IntoIterator for &'_ VerdefSection<'a> {
+        type Item = <VerdefIter<'a> as Iterator>::Item;
+        type IntoIter = VerdefIter<'a>;
+
+        #[inline]
+        fn into_iter(self) -> Self::IntoIter {
+            VerdefIter {
+                bytes: self.bytes,
+                count: self.count,
+                index: 0,
+                offset: 0,
+                ctx: self.ctx,
+            }
+        }
+    }
+
+    /// Iterator over the [`Verdef`] entries from the [`SHT_GNU_VERDEF`] section.
+    pub struct VerdefIter<'a> {
+        bytes: &'a [u8],
+        count: usize,
+        index: usize,
+        offset: usize,
+        ctx: container::Ctx,
+    }
+
+    impl<'a> Iterator for VerdefIter<'a> {
+        type Item = Verdef<'a>;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            if self.index >= self.count {
+                None
+            } else {
+                self.index += 1;
+
+                // Safe to unwrap as the length of the byte slice was validated in VerdefSection::parse.
+                let ElfVerdef {
+                    vd_version,
+                    vd_flags,
+                    vd_ndx,
+                    vd_cnt,
+                    vd_hash,
+                    vd_aux,
+                    vd_next,
+                } = self.bytes.pread_with(self.offset, self.ctx.le).unwrap();
+
+                // Get a slice of bytes of the `Verdaux` entries.
+                //
+                // | Verdef | .. | Verdef | Verdaux | Verdaux | .. | Verdef | ..
+                // ^-------------^
+                //  offset       ^---------^
+                //                vd_aux
+                //               ^---------------------------------^
+                //                 vd_next
+                //
+                // Safe to unwrap as the length of the byte slice was validated in VerdefSection::parse.
+                let len = if vd_next > 0 {
+                    (vd_next - vd_aux) as usize
+                } else {
+                    // For the last entry, ElfVerdef->vd_next == 0.
+                    // Therefore we compute the remaining length of the bytes buffer.
+                    self.bytes.len() - self.offset - vd_aux as usize
+                };
+                let bytes: &'a [u8] = self
+                    .bytes
+                    .pread_with(self.offset + vd_aux as usize, len)
+                    .unwrap();
```

Comment:
For this `unwrap`, and the one above: `VerdefSection::parse` only validated that the section offset and length are within the file size, not whether the section is long enough to contain the entries, and not whether `vd_next` and `vd_aux` are valid. You need to handle invalid values for those.

Similar comments for other unwraps.

Reply:
Removed all calls to unwrap, but this needs another round of review.